File: src/simulation/physics/vpm/biotSavart.js
import { controlNaturalCirculationDirection } from '../runtime/naturalBiotSavartModulation.js'
const FOUR_PI = 4 * Math.PI

function getCoreRadius(particle, params) {
  const minCore = Math.max(params.minCoreRadius ?? 0.01, 1e-4)
  const sigma = particle.coreRadius ?? params.coreRadiusSigma ?? minCore
  return Math.max(minCore, sigma)
}

function getInjectedVelocityContribution(particle, params) {
  const jetDuration = Math.max(params.pulseDuration ?? 0, 1e-4)
  const age = Math.max(particle.age ?? 0, 0)

  if (age >= jetDuration) {
    return { x: 0, y: 0, z: 0 }
  }

  const injectVx = particle.injectVx ?? particle.vx ?? 0
  const injectVy = particle.injectVy ?? particle.vy ?? 0
  const injectVz = particle.injectVz ?? particle.vz ?? 0
  const weight = 1 - age / jetDuration

  return {
    x: injectVx * weight,
    y: injectVy * weight,
    z: injectVz * weight,
  }
}

export function computeVelocityBiotSavart(particles, params) {
  const count = particles.length
  const eps = 1e-8
  const interactionRadius = Math.max(params.interactionRadius ?? 0, 0)
  const interactionRadius2 = interactionRadius * interactionRadius

  for (let i = 0; i < count; i += 1) {
    const particle = particles[i]
    const injectedVelocity = getInjectedVelocityContribution(particle, params)
    let vx = injectedVelocity.x
    let vy = injectedVelocity.y
    let vz = injectedVelocity.z

    if (params.useBiotSavart) {
      for (let j = 0; j < count; j += 1) {
        if (i === j) {
          continue
        }

        const source = particles[j]
        const rx = particle.x - source.x
        const ry = particle.y - source.y
        const rz = particle.z - source.z
        const r2 = rx * rx + ry * ry + rz * rz
        if (interactionRadius > 0 && r2 > interactionRadius2) {
          continue
        }
        const sigma = getCoreRadius(source, params)
        const denom = (r2 + sigma * sigma) ** 1.5

        if (denom <= eps) {
          continue
        }

        const omega = controlNaturalCirculationDirection(
          source,
          source.vorticity ?? { x: 0, y: 0, z: 0 },
          params,
        )
        const cx = ry * omega.z - rz * omega.y
        const cy = rz * omega.x - rx * omega.z
        const cz = rx * omega.y - ry * omega.x
        const gamma = source.gamma ?? params.gamma ?? 0
        const factor = gamma / (FOUR_PI * denom)

        vx += cx * factor
        vy += cy * factor
        vz += cz * factor
      }
    } else {
      vx += (particle.vx ?? 0) / Math.max(params.timeScale ?? 1, 1e-4)
      vy += (particle.vy ?? 0) / Math.max(params.timeScale ?? 1, 1e-4)
      vz += (particle.vz ?? 0) / Math.max(params.timeScale ?? 1, 1e-4)
    }

    particle.flowVx = vx
    particle.flowVy = vy
    particle.flowVz = vz
  }
}

File: torus-core/crates/torus-physics/src/biot_savart.rs
//! Biot-Savart velocity computation — direct O(N²).
//!
//! Port of `src/simulation/physics/vpm/biotSavart.js`.
//! v(x_i) = Σ_j (γ_j / 4π) · (r × ω_j) / (|r|² + σ²)^(3/2)

use glam::DVec3;
use rayon::prelude::*;

use crate::particle::Particle;
use crate::params::SimParams;

const FOUR_PI: f64 = 4.0 * std::f64::consts::PI;

fn get_core_radius(particle: &Particle, params: &SimParams) -> f64 {
    let min_core = params.min_core_radius.max(1e-4);
    particle.core_radius.max(min_core)
}

fn injected_velocity_contribution(particle: &Particle, params: &SimParams) -> DVec3 {
    let jet_duration = params.pulse_duration.max(1e-4);
    let age = particle.age.max(0.0);

    if age >= jet_duration {
        return DVec3::ZERO;
    }

    let weight = 1.0 - age / jet_duration;
    particle.inject_velocity * weight
}

/// Direct Biot-Savart: O(N²), parallelized with rayon.
///
/// Updates `particle.flow_velocity` for each particle.
pub fn compute_velocity_biot_savart(particles: &mut [Particle], params: &SimParams) {
    let count = particles.len();
    if count == 0 {
        return;
    }

    let interaction_radius2 = if params.interaction_radius > 0.0 {
        params.interaction_radius * params.interaction_radius
    } else {
        0.0
    };
    let use_interaction_cutoff = params.interaction_radius > 0.0;

    // Snapshot immutable data for parallel read
    let positions: Vec<DVec3> = particles.iter().map(|p| p.position).collect();
    let vorticities: Vec<DVec3> = particles.iter().map(|p| p.vorticity).collect();
    let gammas: Vec<f64> = particles.iter().map(|p| p.gamma).collect();
    let sigmas: Vec<f64> = particles.iter().map(|p| get_core_radius(p, params)).collect();

    let flows: Vec<DVec3> = (0..count)
        .into_par_iter()
        .map(|i| {
            let inject = injected_velocity_contribution(&particles[i], params);
            let mut v = inject;

            if params.use_biot_savart {
                let pos_i = positions[i];

                for j in 0..count {
                    if i == j {
                        continue;
                    }

                    let r = pos_i - positions[j];
                    let r2 = r.length_squared();

                    if use_interaction_cutoff && r2 > interaction_radius2 {
                        continue;
                    }

                    let sigma = sigmas[j];
                    let denom = (r2 + sigma * sigma).powf(1.5);

                    if denom <= 1e-8 {
                        continue;
                    }

                    let omega = vorticities[j];
                    let cross = r.cross(omega);
                    let gamma = gammas[j];
                    let factor = gamma / (FOUR_PI * denom);

                    v += cross * factor;
                }
            }

            v
        })
        .collect();

    for (i, particle) in particles.iter_mut().enumerate() {
        particle.flow_velocity = flows[i];
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zero_particles_no_panic() {
        let params = SimParams::default();
        let mut particles: Vec<Particle> = vec![];
        compute_velocity_biot_savart(&mut particles, &params);
    }

    #[test]
    fn single_particle_gets_inject_velocity() {
        let params = SimParams {
            pulse_duration: 1.0,
            ..SimParams::default()
        };
        let mut particles = vec![Particle {
            inject_velocity: DVec3::new(1.0, 0.0, 0.0),
            age: 0.0,
            ..Particle::new(0, DVec3::ZERO)
        }];
        compute_velocity_biot_savart(&mut particles, &params);
        assert!((particles[0].flow_velocity.x - 1.0).abs() < 1e-10);
    }

    #[test]
    fn two_particles_induce_velocity() {
        let params = SimParams {
            use_biot_savart: true,
            pulse_duration: 0.0,
            gamma: 1.0,
            min_core_radius: 0.1,
            ..SimParams::default()
        };
        let mut particles = vec![
            Particle {
                gamma: 1.0,
                vorticity: DVec3::new(0.0, 0.0, 1.0),
                core_radius: 0.1,
                age: 100.0,
                ..Particle::new(0, DVec3::new(0.0, 0.0, 0.0))
            },
            Particle {
                gamma: 1.0,
                vorticity: DVec3::new(0.0, 0.0, 1.0),
                core_radius: 0.1,
                age: 100.0,
                ..Particle::new(1, DVec3::new(1.0, 0.0, 0.0))
            },
        ];
        compute_velocity_biot_savart(&mut particles, &params);

        // Particle 0 should have non-zero flow from particle 1
        let flow0 = particles[0].flow_velocity;
        assert!(flow0.length() > 1e-6, "Expected non-zero flow, got {:?}", flow0);
    }

    #[test]
    fn biot_savart_antisymmetric() {
        let params = SimParams {
            use_biot_savart: true,
            pulse_duration: 0.0,
            min_core_radius: 0.1,
            ..SimParams::default()
        };
        let mut particles = vec![
            Particle {
                gamma: 1.0,
                vorticity: DVec3::new(0.0, 0.0, 1.0),
                core_radius: 0.1,
                age: 100.0,
                ..Particle::new(0, DVec3::new(-0.5, 0.0, 0.0))
            },
            Particle {
                gamma: 1.0,
                vorticity: DVec3::new(0.0, 0.0, 1.0),
                core_radius: 0.1,
                age: 100.0,
                ..Particle::new(1, DVec3::new(0.5, 0.0, 0.0))
            },
        ];
        compute_velocity_biot_savart(&mut particles, &params);

        // Flow should be antisymmetric in y for symmetric placement along x
        let vy0 = particles[0].flow_velocity.y;
        let vy1 = particles[1].flow_velocity.y;
        assert!(
            (vy0 + vy1).abs() < 1e-10,
            "Expected antisymmetric vy: {} vs {}",
            vy0,
            vy1
        );
    }
}
